### PR TITLE
Poke multi-byte values directly rather than byte-at-a-time.

### PR DESCRIPTION
**Pull request by guardbotmk3**
This gives a modest performance improvement without compromising safety.

Bump required base version so that we are guaranteed to have fast byteswapping

### CODE REVIEW

**Comment by guardbotmk3** [comment]
This is measurably faster. Let me know how you feel about bumping the required base version.

**Comment by minad** [comment]
* The unaligned access is not a good idea since there are platforms where it is disallowed. You could however provide different versions of the unaligned access helpers, which uses your version on x86 and the bytewise version on other platforms. Maybe call the functions unsafeUnalignedGet16 and unsafeUnalignedPut16 etc?
* ~Bumping base should be in a separate PR maybe.~ Why is it needed? Because of byteSwap?
* Why do you expose fromLE16 etc? I would like to keep the API small.

**Comment by minad** [comment]
But in principle I am positive with this change. How much faster is it? I assume on x86_64. Did you run the haskell-perf benchmark?

**Comment by guardbotmk3** [comment]
I ran the serialization benchmark. It was faster on x86_64, but I didn't get tight numbers.

I've cleaned it up so that we have a clean version for memory aligned architectures.
Even on those architectures, we separate byte pushing from byte disassembly.

I've taken out the exports.

As to base, some of the extensions you have enabled require base versions later than 4.7
I was just being fastidious because I imported bSwap* and they are 4.7+

**Comment by minad** [comment]
Looks better but I don't understand why you use memcpy+alloc. This is definitely worse than reading byte by byte as I have now.

> As to base, some of the extensions you have enabled require base versions later than 4.7

I am ok with updating to newer base. The old constraint was just due to copy and pasting I guess from cereal.

> Even on those architectures, we separate byte pushing from byte disassembly.

What do you mean by that?

> I ran the serialization benchmark. It was faster on x86_64, but I didn't get tight numbers.

While persist aims for very good performance and I also assume that the unaligned access will be faster, I would still like to see some before and after numbers. If this is only like 0.5% faster or so it is not worth the effort.

**Comment by minad** [comment]
Concerning the byte by byte reading - the goal is that everything compiles down to inlined primitives and the corresponding assembly instructions. I don't know if that will happen with the small memcpys. I assume GHC won't inline those.

**Comment by guardbotmk3** [comment]
I fixed and benchmarked the aligned accesses.

Here are the benchmarks for 0.1.1.1 vs direct poke:
There is improvement in BinTree Int serialization ~7%, and Iris
Serialization and Deserialization. Everything else remains unchanged.

I suspect that's because most of it is byte oriented.


0.1.1.1:

benchmarking serialization (time)/[Direction]-persist
time                 1.408 ms   (1.378 ms .. 1.491 ms)
                     0.976 R²   (0.933 R² .. 1.000 R²)
mean                 1.397 ms   (1.378 ms .. 1.473 ms)
std dev              124.6 μs   (997.2 ns .. 265.2 μs)
variance introduced by outliers: 66% (severely inflated)

benchmarking deserialization (time)/[Direction]-persist
time                 7.603 ms   (6.575 ms .. 8.648 ms)
                     0.860 R²   (0.762 R² .. 0.925 R²)
mean                 7.402 ms   (6.834 ms .. 8.491 ms)
std dev              2.001 ms   (1.396 ms .. 3.184 ms)
variance introduced by outliers: 91% (severely inflated)

benchmarking serialization (time)/BinTree Int-persist
time                 59.14 ms   (53.57 ms .. 62.45 ms)
                     0.971 R²   (0.920 R² .. 0.992 R²)
mean                 59.20 ms   (55.98 ms .. 63.08 ms)
std dev              6.815 ms   (4.792 ms .. 9.817 ms)
variance introduced by outliers: 40% (moderately inflated)

benchmarking deserialization (time)/BinTree Int-persist
time                 224.6 ms   (184.8 ms .. 259.2 ms)
                     0.975 R²   (0.875 R² .. 1.000 R²)
mean                 215.7 ms   (203.2 ms .. 230.0 ms)
std dev              18.91 ms   (11.21 ms .. 26.97 ms)
variance introduced by outliers: 16% (moderately inflated)

benchmarking serialization (time)/BinTree Direction-persist
time                 56.51 ms   (53.48 ms .. 62.85 ms)
                     0.977 R²   (0.941 R² .. 1.000 R²)
mean                 56.41 ms   (54.71 ms .. 59.40 ms)
std dev              4.013 ms   (1.332 ms .. 5.889 ms)
variance introduced by outliers: 23% (moderately inflated)

benchmarking deserialization (time)/BinTree Direction-persist
time                 227.5 ms   (196.6 ms .. 260.0 ms)
                     0.992 R²   (0.980 R² .. 1.000 R²)
mean                 213.0 ms   (202.6 ms .. 222.2 ms)
std dev              13.39 ms   (9.204 ms .. 17.72 ms)
variance introduced by outliers: 15% (moderately inflated)

benchmarking serialization (time)/Cars-persist
time                 3.681 ms   (3.255 ms .. 4.144 ms)
                     0.904 R²   (0.840 R² .. 0.962 R²)
mean                 3.576 ms   (3.394 ms .. 3.921 ms)
std dev              799.3 μs   (500.7 μs .. 1.265 ms)
variance introduced by outliers: 91% (severely inflated)

benchmarking deserialization (time)/Cars-persist
time                 8.517 ms   (7.377 ms .. 9.656 ms)
                     0.862 R²   (0.754 R² .. 0.928 R²)
mean                 8.647 ms   (7.981 ms .. 9.682 ms)
std dev              2.261 ms   (1.686 ms .. 3.307 ms)
variance introduced by outliers: 90% (severely inflated)

benchmarking serialization (time)/Iris-persist
time                 4.581 ms   (3.978 ms .. 5.047 ms)
                     0.851 R²   (0.722 R² .. 0.925 R²)
mean                 4.671 ms   (4.170 ms .. 5.655 ms)
std dev              1.940 ms   (1.313 ms .. 3.271 ms)
variance introduced by outliers: 97% (severely inflated)

benchmarking deserialization (time)/Iris-persist
time                 18.54 ms   (16.21 ms .. 20.62 ms)
                     0.928 R²   (0.850 R² .. 0.971 R²)
mean                 18.08 ms   (16.38 ms .. 20.05 ms)
std dev              4.431 ms   (3.064 ms .. 6.856 ms)
variance introduced by outliers: 85% (severely inflated)



With direct peek/poke:

benchmarking serialization (time)/[Direction]-persist
time                 1.388 ms   (1.359 ms .. 1.469 ms)
                     0.976 R²   (0.932 R² .. 1.000 R²)
mean                 1.378 ms   (1.359 ms .. 1.456 ms)
std dev              126.1 μs   (899.9 ns .. 267.8 μs)
variance introduced by outliers: 67% (severely inflated)

benchmarking deserialization (time)/[Direction]-persist
time                 7.636 ms   (6.684 ms .. 8.436 ms)
                     0.877 R²   (0.785 R² .. 0.938 R²)
mean                 7.366 ms   (6.812 ms .. 8.262 ms)
std dev              1.956 ms   (1.371 ms .. 3.067 ms)
variance introduced by outliers: 91% (severely inflated)

benchmarking serialization (time)/BinTree Int-persist
time                 54.91 ms   (48.80 ms .. 58.40 ms)
                     0.966 R²   (0.896 R² .. 0.992 R²)
mean                 54.42 ms   (51.09 ms .. 59.13 ms)
std dev              6.989 ms   (4.951 ms .. 10.01 ms)
variance introduced by outliers: 48% (moderately inflated)

benchmarking deserialization (time)/BinTree Int-persist
time                 221.9 ms   (183.1 ms .. 260.5 ms)
                     0.975 R²   (0.866 R² .. 1.000 R²)
mean                 213.6 ms   (199.8 ms .. 227.4 ms)
std dev              18.87 ms   (11.46 ms .. 26.62 ms)
variance introduced by outliers: 16% (moderately inflated)

benchmarking serialization (time)/BinTree Direction-persist
time                 55.98 ms   (52.95 ms .. 62.55 ms)
                     0.976 R²   (0.935 R² .. 1.000 R²)
mean                 55.91 ms   (54.22 ms .. 59.25 ms)
std dev              4.025 ms   (1.309 ms .. 5.925 ms)
variance introduced by outliers: 23% (moderately inflated)

benchmarking deserialization (time)/BinTree Direction-persist
time                 229.7 ms   (198.1 ms .. 263.3 ms)
                     0.991 R²   (0.980 R² .. 1.000 R²)
mean                 214.1 ms   (205.0 ms .. 224.1 ms)
std dev              13.53 ms   (9.917 ms .. 18.35 ms)
variance introduced by outliers: 15% (moderately inflated)

benchmarking serialization (time)/Cars-persist
time                 3.621 ms   (3.214 ms .. 4.067 ms)
                     0.904 R²   (0.841 R² .. 0.961 R²)
mean                 3.542 ms   (3.376 ms .. 3.944 ms)
std dev              784.7 μs   (496.9 μs .. 1.270 ms)
variance introduced by outliers: 91% (severely inflated)

benchmarking deserialization (time)/Cars-persist
time                 8.588 ms   (7.537 ms .. 9.557 ms)
                     0.876 R²   (0.782 R² .. 0.936 R²)
mean                 8.472 ms   (7.750 ms .. 9.472 ms)
std dev              2.228 ms   (1.569 ms .. 3.382 ms)
variance introduced by outliers: 90% (severely inflated)

benchmarking serialization (time)/Iris-persist
time                 3.628 ms   (3.259 ms .. 3.975 ms)
                     0.820 R²   (0.686 R² .. 0.912 R²)
mean                 3.967 ms   (3.473 ms .. 4.724 ms)
std dev              1.923 ms   (1.348 ms .. 2.954 ms)
variance introduced by outliers: 98% (severely inflated)

benchmarking deserialization (time)/Iris-persist
time                 16.92 ms   (14.84 ms .. 18.90 ms)
                     0.922 R²   (0.842 R² .. 0.968 R²)
mean                 16.87 ms   (15.39 ms .. 18.97 ms)
std dev              4.298 ms   (2.965 ms .. 6.516 ms)
variance introduced by outliers: 86% (severely inflated)

File: persist.cabal [hunk 1]
```diff
@@ -25,13 +25,20 @@ source-repository head
...
         default-language:       Haskell2010
 
-        build-depends:          base >= 4.4 && < 5, containers,
+        build-depends:          base >= 4.7 && < 5, containers,
                                 bytestring >= 0.10.4 && < 1,
                                 text >= 1.2 && < 1.3
 
+        if flag(force-alignment) || arch(PPC) || arch(PPC64) || arch(Mips) || arch(Sparc) || arch(Arm)
+                cpp-options: -DALIGNED_MEMORY
+
```
Comment (minad) [inline persist.cabal:46]: I would prefer whitelisting of unaligned access. `flag force-unaligned` and then a whitelist.

Comment (guardbotmk3) [reply persist.cabal:46]: Will do. Probably tomorrow.

**Comment by minad** [comment]
looks good besides my one comment :+1:

**Comment by minad** [comment]
concerning the test suite, we should also run the tests once with force-unaligned=true and once with force-aligned=true.

**Comment by guardbotmk3** [comment]
OK, This should be in a state where you're happy with it. I changed the flags and .travis.yml

**Comment by minad** [comment]
Merged with minor nitpicks, for some reason 8.6.1 is failing. I have to check this.

**Pull request closed without merging**